### PR TITLE
 Refactoring Server Job related code + removing obsoleted reservation job

### PR DESCRIPTION
**Pull request by agrawalravi90**
#### Describe Bug or Feature

This is the first in a series of refactoring PRs. The aim is to clean up the server code, make it more readable and easier to develop. The ultimate goal is to go towards object oriented design and use a language like C++ instead of C. This is a small first step in that direction, focused on the server's job object and related code.

#### Describe Your Change

Did the following:
- The functions which only "consume" the job object shouldn't modify its data, so added 'const' to such functions' job parameter, this opened a can of worms, had to modify a ton of functions to add const, but this should be worth it, it will help us ensure that all functions which modify the job data stay in one place only
    - This is again being done in steps, I've not modified all such functions yet.
- Also removed a legacy, outdated feature called "reservation job", this needed removal of multiple job attributes and other related code, along with the reservation attribute "reserve_type"

### CODE REVIEW

File: src/lib/Libattr/attr_func.c [hunk 1]
```diff
@@ -1584,3 +1584,24 @@ void free_attrl_list(struct attrl *at_list)
 
 }
 
+/**
+ * @brief	Helper function to remove a flag from an array of attributes
+ *
+ * @param[in]	pattr - pointer to the attribute list
+ * @param[in]	flag - the flag to unset
```
Comment (bhroam) [inline src/lib/Libattr/attr_func.c:1591]: There is nothing stopping you from passing in more than one flag to use this function, I'd rename this variable 'flags'

File: src/lib/Libdb/db_postgres_resv.c [hunk 2]
```diff
@@ -84,29 +83,28 @@ pg_db_prepare_resv_sqls(pbs_db_conn_t *conn)
...
-		"ri_etime = $7, "
-		"ri_duration = $8, "
-		"ri_tactive = $9, "
-		"ri_svrflags = $10, "
-		"ri_numattr = $11, "
-		"ri_resvTag = $12, "
-		"ri_un_type = $13, "
-		"ri_fromsock = $14, "
-		"ri_fromaddr = $15, "
+		"ri_stime = $5, "
```
Comment (bhroam) [inline src/lib/Libdb/db_postgres_resv.c:94]: How will this affect upgrades?  This affects the schema, so when the schema changes, will the database as well?

File: src/scheduler/resv_info.c [hunk 3]
```diff
@@ -272,291 +272,288 @@ query_reservations(server_info *sinfo, struct batch_status *resvs)
...
-									while (req != NULL) {
-										if (req->type.is_consumable) {
-											res = find_resource(ns->ninfo->res, req->def);
-											if (res != NULL)
-												res->assigned += req->amount;
-										}
-										req = req->next;
+					if (rjob->job->is_running) {
+						/* the reservations resv_nodes is pointing to
+							* a node_info array with just the reservations part of the node
```
Comment (bhroam) [inline src/scheduler/resv_info.c:311]: Are the *'s not lining up?  (here and other places)

File: src/server/accounting.c [hunk 4]
```diff
@@ -1084,36 +1069,11 @@ account_record(int acctype, job *pjob, char *text)
  *
  * @return	void
  */
-void
-account_recordResv(int acctype, resc_resv *presv, char *text)
+void account_recordResv(int acctype, resc_resv *presv, char *text)
```
Comment (bhroam) [inline src/server/accounting.c:1072]: Put the void back on the line above the function

File: src/server/accounting.c [hunk 5]
```diff
@@ -1164,21 +1124,25 @@ account_jobstr2(job *pjob, int type)
 
 		nd = 0; /* compute total size needed in buf */
 		pal = GET_NEXT(attrlist);
-		while (pal != NULL) {
+		while (pal != NULL)
+		{
```
Comment (bhroam) [inline src/server/accounting.c:1128]: Put the { back up on the line with the while().  It looks like a lot of {'s in this function were moved.  Move them back.

File: src/server/svr_jobfunc.c [hunk 6]
```diff
@@ -577,10 +577,7 @@ svr_dequejob(job *pjob)
 		correct_ct(pque);
 #endif	/* NDEBUG */
 
-	pjob->ji_wattr[(int)JOB_ATR_qtime].at_flags &= ~ATR_VFLAG_SET;
-
```
Comment (bhroam) [inline src/server/svr_jobfunc.c:581]: Why remove this?

File: src/server/accounting.c (hunk 5 above)
Comment (agrawalravi90) [reply src/server/accounting.c:1128]: I'm suffering from:
https://github.com/Microsoft/vscode/issues/49662 :(

Will reformat it, thanks

File: src/lib/Libdb/db_postgres_resv.c (hunk 2 above)
Comment (agrawalravi90) [reply src/lib/Libdb/db_postgres_resv.c:94]: I hadn't thought about that. You are correct, this changes the schema, I'm not sure how upgrades will work. Do you know how we usually handle this?

Comment (bhroam) [reply src/lib/Libdb/db_postgres_resv.c:94]: I think habitat updates the schema and the tables in the database.  I'm not exactly sure how it is done.  I suspect @subhasisb will know.

**Comment by agrawalravi90** [comment]
@bhroam  I've made changes to make overlay upgrades possible with the schema change. Here's the results of my testing:

**Before:**
```
pbs_datastore=# \d pbs.resv
                            Table "pbs.resv"
   Column    |            Type             |          Modifiers          
-------------+-----------------------------+-----------------------------
 ri_resvid   | text                        | not null
 ri_queue    | text                        | not null
 ri_state    | integer                     | not null
 ri_substate | integer                     | not null
 ri_type     | integer                     | not null
 ri_stime    | bigint                      | not null
 ri_etime    | bigint                      | not null
 ri_duration | bigint                      | not null
 ri_tactive  | integer                     | not null
 ri_svrflags | integer                     | not null
 ri_numattr  | integer                     | not null
 ri_resvtag  | integer                     | not null
 ri_un_type  | integer                     | not null
 ri_fromsock | integer                     | not null
 ri_fromaddr | bigint                      | not null
 ri_savetm   | timestamp without time zone | not null
 ri_creattm  | timestamp without time zone | not null
 attributes  | hstore                      | not null default ''::hstore
Indexes:
    "resv_pk" PRIMARY KEY, btree (ri_resvid)
pbs_datastore=# select pbs_schema_version from pbs.info;
 pbs_schema_version 
--------------------
 1.4.0
(1 row)
```

**After (notice that ri_type is gone and schema version number has changed):**
```
pbs_datastore=# \d pbs.resv
                            Table "pbs.resv"
   Column    |            Type             |          Modifiers          
-------------+-----------------------------+-----------------------------
 ri_resvid   | text                        | not null
 ri_queue    | text                        | not null
 ri_state    | integer                     | not null
 ri_substate | integer                     | not null
 ri_stime    | bigint                      | not null
 ri_etime    | bigint                      | not null
 ri_duration | bigint                      | not null
 ri_tactive  | integer                     | not null
 ri_svrflags | integer                     | not null
 ri_numattr  | integer                     | not null
 ri_resvtag  | integer                     | not null
 ri_un_type  | integer                     | not null
 ri_fromsock | integer                     | not null
 ri_fromaddr | bigint                      | not null
 ri_savetm   | timestamp without time zone | not null
 ri_creattm  | timestamp without time zone | not null
 attributes  | hstore                      | not null default ''::hstore
Indexes:
    "resv_pk" PRIMARY KEY, btree (ri_resvid)

pbs_datastore=# 

pbs_datastore=# select pbs_schema_version from pbs.info;
 pbs_schema_version 
--------------------
 1.5.0
(1 row)
```

And this was done without having to remove PBS_HOME, just had to run pbs_habitat script to upgrade the schema. Please review the latest commits, thanks!

**Comment by agrawalravi90** [comment]
@bhroam gentle reminder for this, whenever you get some time, thanks.

File: src/include/attribute.h [hunk 7]
```diff
@@ -192,7 +192,7 @@ typedef struct attribute attribute;
 struct attribute_def {
 	char	*at_name;
 	int	(*at_decode)(attribute *patr, char *name, char *rn, char *val);
-	int	(*at_encode)(attribute *pattr, pbs_list_head *phead,
+	int	(*at_encode)(const attribute *pattr, pbs_list_head *phead,
```
Comment (hirenvadalia) [inline src/include/attribute.h:195]: Please put all argument in one line (applies to other files also)

File: src/scheduler/resv_info.c [hunk 8]
```diff
@@ -272,291 +272,288 @@ query_reservations(server_info *sinfo, struct batch_status *resvs)
...
-			}
-			resresv->resv->resv_queue =
-				find_queue_info(sinfo->queues, resresv->resv->queuename);
-			if (is_resresv_running(resresv)) {
-				for (j = 0; resresv->ninfo_arr[j] != NULL; j++)
-					resresv->ninfo_arr[j]->num_run_resv++;
-			}
+		if (resresv->node_set_str != NULL) {
+			resresv->node_set = create_node_array_from_str(
+				resresv->server->unassoc_nodes, resresv->node_set_str);
```
Comment (hirenvadalia) [inline src/scheduler/resv_info.c:277]: Please merge this line with above

File: src/scheduler/resv_info.c [hunk 9]
```diff
@@ -272,291 +272,288 @@ query_reservations(server_info *sinfo, struct batch_status *resvs)
...
+		}
+		resresv->resv->resv_queue =
+			find_queue_info(sinfo->queues, resresv->resv->queuename);
```
Comment (hirenvadalia) [inline src/scheduler/resv_info.c:280]: same as above

File: src/scheduler/resv_info.c [hunk 10]
```diff
@@ -272,291 +272,288 @@ query_reservations(server_info *sinfo, struct batch_status *resvs)
...
+									req = req->next;
 								}
-								else {
+							}
+							else {
 #ifdef NAS /* localmod 031 */
-									log_eventf(PBSEVENT_RESV, PBS_EVENTCLASS_RESV, LOG_INFO, rjob->name, 
-										"Job has been assigned a node that doesn't exist in its reservation: %s", ns->ninfo->name);
+								log_eventf(PBSEVENT_RESV, PBS_EVENTCLASS_RESV, LOG_INFO, rjob->name,
+									"Job has been assigned a node that doesn't exist in its reservation: %s", ns->ninfo->name);
```
Comment (hirenvadalia) [inline src/scheduler/resv_info.c:340]: Add one more tab to this line to align properly

File: src/server/job_func.c [hunk 11]
```diff
@@ -1992,295 +1954,6 @@ job_or_resv_init_wattr(void *pobj, int obj_type)
...
-	 *This avoids a timing issue where "stat" requests coming from
-	 *the scheduler and the "req_commit" request can be such that
-	 *the scheduler sees the reservation before the reservation's
-	 *job is on svr_alljobs.  Immediate deleting of such a reservation
-	 *ends up deleting the reservation but leaves the job structure
-	 *still in existence (on the svr_alljobs list) following the
-	 *req_commit request).
-	 */
-
-	append_link(&svr_newresvs, &presv->ri_allresvs, presv);
```
Comment (hirenvadalia) [inline src/server/job_func.c:2150]: Please remove `svr_newresvs` global list, as it was only used for "reservation job" only

File: src/cmds/scripts/pbs_habitat.in [hunk 12]
```diff
@@ -658,28 +652,38 @@ if [ "${PBS_START_SERVER:-0}" != 0 ] ; then
...
+		if [ "$sys_pgsql_ver" != "$old_pgsql_ver" ]; then
+			# Upgrade postgres
+			upgrade_pbs_database $sys_pgsql_ver $old_pgsql_ver
+			ret=$?
+			if [ $ret -ne 0 ]; then
+				if [ $ret -eq 2 ]; then
+					echo "It appears that PostgreSQL has been upgraded independently of PBS Pro."
+					echo "The PBS Pro database must be manually upgraded. Please refer to the"
+					echo "instructions in the release notes for version @PBS_VERSION@ located here:"
+					echo "https://github.com/PBSPro/pbspro/releases"
```
Comment (bhroam) [inline src/cmds/scripts/pbs_habitat.in:671]: This is about to change to openpbs.  Since that PR is going in next, talk with Hiren for what this should be.

Comment (hirenvadalia) [reply src/cmds/scripts/pbs_habitat.in:671]: @agrawalravi90 As my renaming PR is merged, I think you will get the new text here, but let me know if you need any help.

File: src/include/attribute.h (hunk 7 above)
Comment (agrawalravi90) [reply src/include/attribute.h:195]: it'll become a very long line, i think breaking is more readable than scrolling left-right in your IDE to see all parameters. So, I'd like to leave it as is for now. I will attempt to reduce the number of parameters for functions which take too many in future refactoring PRs.

File: src/server/job_func.c (hunk 11 above)
Comment (agrawalravi90) [reply src/server/job_func.c:2150]: Thanks for catching that!

File: src/cmds/scripts/pbs_habitat.in (hunk 12 above)
Comment (agrawalravi90) [reply src/cmds/scripts/pbs_habitat.in:671]: Thanks, I'll take care of it when I resolve conflicts

**Comment by agrawalravi90** [comment]
@hirenvadalia  and @bhroam thanks guys, ive addressed your comments, please provide further feedback.

**Comment by agrawalravi90** [comment]
@bhroam  and @hirenvadalia  Thanks for sign-offs, I've resolved conflicts and rebased, ran regression and it passed. Please verify that this still looks good, thanks!

**Comment by hirenvadalia** [comment]
Thanks @agrawalravi90, changes still look good.

**Comment by agrawalravi90** [comment]
@bhroam  and @hirenvadalia thanks a ton for the review, I've resolved conflicts and ran successful regression, I think this is ready for merge.

**Pull request merged**